include/coreinit/performancemonitor.h

```diff
@@ -0,0 +1,206 @@
+    PM_PMC3_L2_CACHE_OPERATIONS                = 0b01101u << 27,
+    PM_PMC3_L1_LOAD_MISS_CYCLES                = 0b01111u << 27,
+    PM_PMC3_SECOND_SPECULATIVE_BRANCH_RESOLVES = 0b10000u << 27,
+    PM_PMC3_BPU_STALL_LR_CR_CYCLES             = 0b10001u << 27,
+    PM_PMC3_L1_MODIFIED_INTERVENTIONS          = 0b10010u << 27,
+    PM_PMC3_ICBI_SNOOPS                        = 0b10011u << 27,
+    PM_PMC3_CIU_ADDRESS_ONLY_REQUESTS          = 0b10100u << 27,
+    PM_PMC3_BIU_LOAD_REQUESTS                  = 0b10101u << 27,
+    PM_PMC3_CIU_SHARED_INTERVENTIONS           = 0b10110u << 27,
+
+    PM_PMC4_CURRENT                           = 0b00000u << 22,
+    PM_PMC4_CPU_CYCLES                        = 0b00001u << 22,
+    PM_PMC4_INSTRUCTIONS_COMPLETED            = 0b00010u << 22,
+    PM_PMC4_TBL_RISING_TRANSITIONS            = 0b00011u << 22,
+    PM_PMC4_INSTRUCTIONS_DISPATCHED           = 0b00100u << 22,
+    PM_PMC4_L2_CASTOUTS                       = 0b00101u << 22,
+    PM_PMC4_DTLB_SEARCH_CYCLES                = 0b00110u << 22,
+    PM_PMC4_BRANCHES_MISPREDICTED             = 0b01000u << 22,
+    PM_PMC4_INTACT_COND_STORES_COMPLETED      = 0b01010u << 22,
+    PM_PMC4_SYNC_INSTRUCTIONS_COMPLETED       = 0b01011u << 22,
+    PM_PMC4_SNOOP_RETRIES                     = 0b01100u << 22,
+    PM_PMC4_INTEGER_OPERATIONS                = 0b01101u << 22,
+    PM_PMC4_BPU_STALL_TWO_BRANCHES_CYCLES     = 0b01110u << 22,
+    PM_PMC4_L2_MODIFIED_INTERVENTIONS         = 0b10000u << 22,
+    PM_PMC4_TLBIE_SNOOPS                      = 0b10001u << 22,
+    PM_PMC4_L2_BANK_REFRESH_OVERFLOWS         = 0b10010u << 22,
+    PM_PMC4_CIU_ARTRY_COUNT                   = 0b10100u << 22,
+    PM_PMC4_BIU_STORE_REQUESTS                = 0b10101u << 22,
+    PM_PMC4_CIU_TWO_CORE_SHARED_INTERVENTIONS = 0b10110u << 22,
+} PerfMonMMCR1Flags;
+
+/**
+ * Write to performance monitor registers.
+ *
+ * Performance monitor registers can only be written by the kernel, this allows userspace
+ * to write to them.
+ *
+ * \param arg_mask OR-ed values from `PerfMonArg`, indicating which of the following
+ * arguments are to be written to registers.
+ *
+ * \param mmcr0 OR-ed values from `PerfMonMMCR0Flags` to write to register MMCR0.
+ * \param mmcr1 OR-ed values from `PerfMonMMCR1Flags` to write to register MMCR1.
+ * \param pmc1 Value to write to register PMC1.
+ * \param pmc2 Value to write to register PMC2.
+ * \param pmc3 Value to write to register PMC3.
+ * \param pmc4 Value to write to register PMC4.
+ */
+void OSSetPerformanceMonitor(uint32_t arg_mask,
+                             uint32_t mmcr0,
+                             uint32_t mmcr1,
+                             uint32_t pmc1,
+                             uint32_t pmc2,
+                             uint32_t pmc3,
+                             uint32_t pmc4);
+
+/**
+ * Convenience function to read from UPMC1.
+ */
+static inline
+uint32_t
+__wut_read_upmc1()
```

Comment:
The `__wut_` prefix should be used for wut internal functions which aren't meant to be exposed. Maybe something like `OSReadUPMC1` or `OSGetUPMC1` could work?